examples/publisher.js

```diff
@@ -1,21 +1,26 @@
-var Publisher = require('../').Publisher;
+let Publisher = require('../').Publisher;
 
 // Instantiate a new Publisher component.
-var randomPublisher = new Publisher({
+let randomPublisher = new Publisher({
     name: 'randomPub',
     // namespace: 'rnd',
-    broadcasts: ['randomUpdate']
+    broadcasts: ['*'],
```

Comment:
this should be `['randomUpdate1', '#room1::randomUpdate', '#room2::randomUpdate']`. We want to be as explicit as possible, as this is also a filter for the Sockend component (messages not listed here will not be published to the frontend, but still can be published on the backend).

Reply:
ahh, but being explicit in this case is problematic because the publishing client might not know what rooms he wants to publish to or even what rooms might exist at any point in time. rooms could be very dynamic in different scenarios. for that reason i preferred to leave it open and only specify the actual topic. 

with that in mind we had 2 solutions so far, each with it's own drawbacks:
1. the subscribesTo['*'] solution, with the drawbacks you mentioned above.
1. the previous wrapper based solution, with it's own drawbacks of nonstandard message format.

if we opt for the explicit approach, we need to provide an elegant solution for publishers that might be sending events to an unknown and always dynamic list of rooms. but seems to me like that would require even deeper changes and would end up being a less elegant api. 

if both the above solutions are unacceptable i guess it's back to the drawing board :)

Reply:
We could also claim that it's not the publisher's responsibility to decide the room; since cote is primarily a microservices framework for the backend, and the extension to the frontend is an add-on and not the primary goal. Since the rooms don't make sense for the backend implementations, we can say that this feature addition to the publishers are not according to the main purpose.

Another solution could be then to include something like a special room redirection middleware for a Sockend so that when it receives a published message, you can plug in your own middleware and decide which room it will be redirected to. What do you think?

Reply:
As an example, the Sockend has the `requesterTransformators`, which we use to plug in for socket.io authentication while the sockend forwards the requests to the responders. This way, if there's an authenticated user in the Sockend, we can modify the incoming request and add user data, so that the backend has the user's session, for example.

Reply:
Yet another solution, a very direct and obvious one, is to modify the `data` that's published, and put a `__room` key in there. The docs can read like; `if you want to make use of rooms, publish an object with a "__room" key`. This saves us from using a wrapper object, and comes with the downside that the users cannot publish simple values but only objects when they want to use rooms. Which is an OK tradeoff. In this case we don't need to do a middleware implementation, we can just look into the incoming `data` object and see if it has a `__room`; if so, we just emit it to that room.

Reply:
hey, you work late :) where are u in the world?
anyway, starting to look into the suggestions.

first this one: `broadcasts is ['*::randomUpdate']`
it would be great if it worked, but the side effect of this is that the `this.event` you get in the subscriber is just the room portion of the wildcard expression. the defined behaviour of EventEmitter2 i guess. I think I tried this in the beginning when I just started working on this feature.

in terms of rooms being in line with the primary goal, i guess you're right but it's a nice addition and a necessary feature in any implementation which aims to segment events to groups of users.

middleware is an option, but makes the implementation for the developer much heavier than need be.

adding __room to the message is basically creating a wrapper. in this case i think it would make sense to return to internal wrappers. it would have the advantage of keeping the external usage clean and simple, plus allowing the user to send simple values not only objects. in this case the wrappers are only used in the leg that transfers from the publisher to the subscriber, and are only visible internally so user is not exposed to it.

in my fantasy as an end user (who really needs this feature), i would like to be able to specify a publisher with `broadcasts['topic1','topic2',....,'topic20']` and then do `publisher.publish('topic1@room1','some value or object')` or `publisher.publish('topic20@room78','some value or object')` and have cote magically handle it for me. 

maximum flexibility with minimum implementation (or configuration). 

i think internal wrappers would serve that purpose well while not bending or going around EventEmitter2 behaviour, and not affecting performance in any noticeable way. we could even have the system only use wrappers when a room is specified, otherwise operate normally.

let me push a version with wrappers in a separate PR so we can evaluate it separately.  i think we might find it has more pros than cons.

Reply:
I'm in Berlin :) I know `this.event` is broken, but I made a prototype that restores it to the original version, so instead of getting only the room portion you get everything.

The problem with `*::randomUpdate` is that it doesn't tell you anything about the channel; it doesn't even have to be a channel. `foo::randomUpdate` is also valid, and this is practically the same as `*`, which is a leak by definition. We could use `#*::randomUpdate` but it's very ugly.

In fact, the approach of string manipulation (like adding `@room` or `#room`) is not ideal because it prevents one from using enums in this case.

On the other hand, there currently is no wrappers in cote, we already inject `type` attribute for Sockend (so we can inject more), and wrapper objects is a design that I don't want to introduce at this point. Telling the user to add `__room` to their published objects is an adequate compromise — if you want to publish to a room, use an object with a field __room in it, and that's it. The only code change is again only in Sockend where it redistributes the message according to the `__room` attribute.

Reply:
ok my friend, i already created a PR with wrappers #103 but feel free to ignore it. I'll just state again that it does have the advantage of allowing simple values whereas __room in the message forces usage of an object.

did not understand this: `because it prevents one from using enums in this case`. how would you use enums?

in any case, i'll make another PR which implements __room in the message. see if you like it.

Reply:
solution with __rooms in PR #104

Reply:
```js
const Messages = {
    UPDATE_1: 'UPDATE_1',
    UPDATE_2: 'UPDATE_2'
};

publisher.publish(Messages.UPDATE_1);
// with the rooms it will be:
publisher.publish(`${Messages.UPDATE_1}@${room}`);
```